src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,997 @@
+      lo[j] == 1                (according to 2.b.2 and 2.5, lo[j] == v[j] == 1 because j < i)
+      r[x]  == lo[x], for x < j (according to 2.b.3 and 2.5, lo[x] == v[x] == r[x] with x < j < i)
+      bits at x > j have lower significance, and are thus irrelevant
+
+      Which leads to r < lo, which contradicts that lo < r (acording to 2.1)
+
+    b.2. If j == i
+      Since r > lo (according to 2.1), there must exist a bit index k such that:
+
+      r[k]  == 1                (2.b.2.1)
+      lo[k] == 0                (2.b.2.2)
+      r[x]  == lo[x], for x < k (2.b.2.3)
+
+      Then, since we have:
+      r[x]  == v[x],  for x < i (according to 2.b.3)
+      v[x]  == lo[x], for x < i (according to 2.5)
+      r[i]  == 0                (according to 2.b.1 because i == j)
+      lo[i] == 0                (according to 2.4)
+
+      this leads to: r[x] == lo[x], for x <= i
+      while r[k] == 1 != lo[k] == 0, we can conclude that k > i
+
+      However, since:
+      lo[x] satisfies bits for 0 <= x < k:
+        According to 2.b.2.3, lo[x] == r[x] and r satisfies bits
+      zeros[k] == 0 (according to 2.b.2.1, r[k] == 1 and r satisfies bits)
+      lo[k]    == 0 (according to 2.b.2.2)
+
+      This contradicts the assumption that i is the largest bit index satisfying such conditions.
+
+    b.3. If j > i
+      ones[j] == v[j] (according to 2.7 since j > i)
+      v[j]    == 1    (according to 2.b.2)
+      r[j]    == 0    (according to 2.b.1)
+
+      This means that r[j] == 0 and ones[j] == 1, this contradicts the assumption that r
+      satisfies bits.
+
+    All cases lead to contradictions, which mean r < v is incorrect, which means that
+    r == v, which means the value v having the above form is the lowest value not smaller
+    than lo that satisfies bits.
+
+  3. Conclusion
+    Our objective now is to find the largest value i that satisfies:
+    - lo[x] satisfies bits for 0 <= x < i
+    - zeros[i] = 0
+    - lo[i] = 0
+
+    Call j the largest value such that lo[x] satisfies bits for 0 <= x < j. This
+    means that j is the smallest value such that lo[j] does not satisfy bits. We
+    call this the first violation. i then can be computed as the largest value
+    <= j such that:
+
+    zeros[i] == lo[i] == 0
+  */
+
+  // The algorithm depends on whether the first violation violates zeros or
+  // ones, if it violates zeros, we have the bit being 1 in zero_violation and
+  // 0 in one_violation. Since all higher bits are 0 in zero_violation and
+  // one_violation, we have zero_violation > one_violation. Similarly, if the
+  // first violation violates ones, we have zero_violation < one_violation.
+  if (zero_violation < one_violation) {
+    // This means that the first bit that does not satisfy the bit requirement
+    // is a 0 that should be a 1. Obviously, since the bit at that position in
+    // ones is 1, the same bit in zeros is 0. Which means this is the value of
+    // i we are looking for.
```

Comment:
```suggestion
    // This means that the first bit that does not satisfy the bit requirement
    // is a 0 that should be a 1. Obviously, since the bit at that position in
    // ones is 1, the same bit in zeros is 0. Which means this is the value of
    // we are looking for.
    // We know i is the largest bit index such that:
    // - lo[x] satisfies bits for 0 <= x < i (2.2)
    // - zeros[i] = 0                        (2.3)
    // - lo[i]    = 0                        (2.4)
    // For the given i, we know that lo satisfies all bits before i, hence (2.2)
    // holds. Further, lo[i] = 0 (2.3), and we have a one violation at i, hence
    // zero[i] = 0 (2.4). Any smaller i would not be the largest possible such
    // index. Any larger i would violate (2.2), since lo[i] does not satisfy bits.
```
I just realized that we should explicitly tie this back in to the proof we wrote above, and link it to the numbers there.

Reply:
Now below, we should try to tie the construction back to (2.5 - 2.7), let me give it a try.